### ISSUE TITLE
crash in raft.sendAppend: requested index at entry is unavailable

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#1191]
This happened in tandem with #1185 and after a flurry of messages

```
W0529 13:18:33.251543       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201
"-"" outside of bounds of range ""-""
```

(which are not surprising in this scenario given that the node never executes any of the splits)

```
I0529 13:18:33.251657       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
W0529 13:18:33.251948       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201"-"" outside of bounds of range ""-""
I0529 13:18:33.251966       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
W0529 13:18:33.253062       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201"-"" outside of bounds of range ""-""
I0529 13:18:33.253138       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
panic: requested entry at index is unavailable

goroutine 64 [running]:
github.com/coreos/etcd/raft.(*raft).sendAppend(0xc2097f4000, 0x300000003)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:250 +0x159
github.com/coreos/etcd/raft.stepLeader(0xc2097f4000, 0x4, 0x200000002, 0x300000003, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:523 +0x6f3
github.com/coreos/etcd/raft.(*raft).Step(0xc2097f4000, 0x4, 0x200000002, 0x300000003, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:487 +0x259
github.com/coreos/etcd/raft.(*multiNode).run(0xc2080ff0e0)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/multinode.go:236 +0xd0b
created by github.com/coreos/etcd/raft.StartMultiNode
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/multinode.go:56 +0xb6
```

**Comment by tschottdorf** [comment]
This is happening reliably with the Put acceptance test, by the way:
`TESTS=TestPut TESTTIMEOUT=24h TESTFLAGS="-d 600s -w" acceptance/run.sh`
after about 40-50k ops.
The first message in the log below is debug added by me in `range_raftstorage.go#Entries`. It likely has to do with the range split (range 12 below resulted from a split announced ~7sec earlier) and the new range winding up without a Raft log. So maybe some race, but it's happening pretty consistently beyond the 40k entries mark.

@bdarnell, any ideas?

>
W0608 14:16:39.880474       1 range_raftstorage.go:106] raft 12: can't load [0,1) with maxBytes 0, found:
[]
panic: requested entry at index is unavailable
goroutine 68 [running]:
github.com/coreos/etcd/raft.(*raft).sendAppend(0xc20807cd80, 0x100000001)
        /go/src/github.com/coreos/etcd/raft/raft.go:250 +0x159
github.com/coreos/etcd/raft.stepLeader(0xc20807cd80, 0x4, 0x300000003, 0x100000001, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/coreos/etcd/raft/raft.go:523 +0x6f3
github.com/coreos/etcd/raft.(*raft).Step(0xc20807cd80, 0x4, 0x300000003, 0x100000001, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/coreos/etcd/raft/raft.go:487 +0x259
github.com/coreos/etcd/raft.(*multiNode).run(0xc2080be300)
        /go/src/github.com/coreos/etcd/raft/multinode.go:236 +0xd0b
created by github.com/coreos/etcd/raft.StartMultiNode
        /go/src/github.com/coreos/etcd/raft/multinode.go:56 +0xb6

**Comment by bdarnell** [comment]
I think this is connected to the "manual split" bug (#1242). My best guess so far is that when a range is created and is waiting for its initial snapshot it's in a weird state, and if it stays in this state long enough for an election timeout it can fail.

**Comment by tschottdorf** [comment]
@bdarnell your SplitRaceTest barfed on one of my branches (which I had just rebased on master, so it definitely has your work already).
https://circleci.com/gh/tschottdorf/cockroach/848
TL;DR:

> I0616 15:23:45.359623     296 range_command.go:835] initiating a split of range=40 ["B018" - "\xff\xff") at key "B019"
I0616 15:23:45.361302     296 range_command.go:772] range 40: new leader lease replica 1:1 00:00:00.000 +1.000s
I0616 15:23:45.377554     296 range_command.go:772] range 41: new leader lease replica 1:1 00:00:00.000 +1.000s
	client_raft_test.go:767: failed to snapshot max range: range_raftstorage.go:257: couldn't find range descriptor

**Comment by tschottdorf** [comment]
more of them, this time on "real" master:
https://circleci.com/gh/cockroachdb/cockroach/3705
https://circleci.com/gh/cockroachdb/cockroach/3707

**Comment by bdarnell** [comment]
I can't reproduce this locally (even with a few busy-loop processes running to emulate the circleci environment), but I've identified the problem. The problem is that `Range.EndTransaction` runs with a batch that will not be committed to rocksdb until after it returns (in `Range.applyRaftCommand`), but the store's in-memory maps are updated during `splitTrigger` (which is called by `EndTransaction`). This leaves a window during which `Store.LookupRange` will return the new range but its range descriptor can't be read from disk (more precisely, the write intent from the split transaction has not been cleared). 

`Store.SplitRange` needs to be delayed until after the batch has been committed. It would be straightforward but ugly to pass a callback from `Range.EndTransaction` (and the triggers) to `applyRaftCommand`. I'll see if I can come up with a cleaner solution.

I'm not sure if an analogous issue exists for merges. In that case we may actually need the opposite, to remove the subsumed range before committing the batch. It may be even better to make it completely atomic: hold the store lock across both the `batch.Commit()` and `Store.SplitRange()`. However, that seems like a very delicate arrangement, and I really don't want to hold the store lock across disk I/O.

**Comment by tschottdorf** [comment]
pretty easy to reproduce:

```
diff --git a/storage/range_command.go b/storage/range_command.go
index ec43d65..7d1f195 100644
--- a/storage/range_command.go
+++ b/storage/range_command.go
@@ -25,6 +25,7 @@ import (
        "fmt"
        "math/rand"
        "sync/atomic"
+       "time"
        "unsafe"

        "github.com/cockroachdb/cockroach/client"
@@ -281,6 +282,7 @@ func (r *Range) EndTransaction(batch engine.Engine, ms *proto.MVCCStats, args *p
                        if ct.SplitTrigger != nil {
                                *ms = proto.MVCCStats{} // clear stats, as split will recompute from scratch.
                                reply.SetGoError(r.splitTrigger(batch, ct.SplitTrigger))
+                               time.Sleep(100 * time.Millisecond)
                        } else if ct.MergeTrigger != nil {
                                *ms = proto.MVCCStats{} // clear stats, as merge will recompute from scratch.
                                reply.SetGoError(r.mergeTrigger(batch, ct.MergeTrigger))
```

**Comment by tschottdorf** [comment]
we could add `Defer(fn() error)` to the `Engine` interface, with the semantics that the callbacks are executed (in the usual reverse order) when `Commit()` is invoked and returns without an error.
Then we don't need to pass callbacks around, and if we lock the range around the Commit() call, we're atomic. But that may be confusing.

Alternatively (and this seems like a clean solution to me), couldn't we move the whole triggers business to where the batch is about to commit in `applyRaftCommand`? After all, that's when those updates should actually happen if they're trying to be atomic.

**Comment by bdarnell** [comment]
We can't move the triggers as they are to `applyRaftCommand`: part of the trigger writes to the DB and needs to be in the same batch as the `EndTransaction`, and part of it modifies in-memory state and needs to be after the batch commits. We could have two sets of trigger methods but that just gets messy.

Adding `Defer()` to `Engine` sounds like a reasonable approach to me.
